src/library/autodj/autodjfeature.cpp

```diff
@@ -164,98 +161,66 @@ bool AutoDJFeature::dragMoveAccept(QUrl url) {
-    // If this newly-added crate is in the auto-DJ queue, add it to the list.
-    if (m_crateDao.isCrateInAutoDj(crateId)) {
-        slotCrateAutoDjChanged(crateId, true);
+    Crate crate;
+    if (m_pTrackCollection->crates().readCrateByName(crateName, &crate) && crate.isAutoDjSource()) {
+        crate.setAutoDjSource(false);
+        m_pTrackCollection->updateCrate(crate);
     }
 }
 
-// Signaled by the crate DAO when a crate is renamed.
-void AutoDJFeature::slotCrateRenamed(int crateId, QString newName) {
-    // Look for this crate ID in our list.  It's OK if it's not found.
-    for (int i = 0; i < m_crateList.length(); ++i) {
-        if (m_crateList[i].first == crateId) {
-            // Change the name of this crate.
-            m_crateList[i].second = newName;
-
-            // Update the display of this crate's name.
-            QModelIndex oCratesIndex = m_childModel.index(0, 0);
-            QModelIndex oCrateIndex = oCratesIndex.child(i, 0);
-            m_childModel.setData(oCrateIndex, QVariant(newName),
-                                 Qt::DisplayRole);
-            break;
+void AutoDJFeature::slotCrateChanged(CrateId crateId) {
+    Crate crate;
+    if (m_pTrackCollection->crates().readCrateById(CrateId(crateId), &crate) && crate.isAutoDjSource()) {
+        for (int i = 0; i < m_crateList.length(); ++i) {
+            if (m_crateList[i].getId() == crateId) {
+                QModelIndex parentIndex = m_childModel.index(0, 0);
+                QModelIndex childIndex = parentIndex.child(i, 0);
+                m_childModel.setData(childIndex, crate.getName(), Qt::DisplayRole);
+                m_crateList[i] = std::move(crate);
+                return; // early exit
+            }
         }
-    }
-}
-
-void AutoDJFeature::slotCrateDeleted(int crateId) {
-    // The crate can't be queried for its auto-DJ status, because it's been
-    // deleted by the time this code is reached.  But we can handle that.
-    // Another solution would be to add a "crateDeleting" signal to CrateDAO.
-    slotCrateAutoDjChanged(crateId, false);
-}
-
-void AutoDJFeature::slotCrateAutoDjChanged(int crateId, bool added) {
-    if (added) {
-        // Get the name of the crate being added to the auto-DJ list.
-        QString strName = m_crateDao.crateName(crateId);
-
-        // Get the index of the row where this crate will be inserted into the
-        // tree.
-        int iRowIndex = m_crateList.length();
-
-        // Add our record of this crate-ID and name.
-        m_crateList.append(qMakePair(crateId, strName));
-
-        // Prepare to add it to the "Crates" tree-item.
-        QList<TreeItem*> lstItems;
-        lstItems.append(new TreeItem(this, strName));
-
-        // Add it to the "Crates" tree-item.
-        QModelIndex oCratesIndex = m_childModel.index(0, 0);
-        m_childModel.insertRows(lstItems, iRowIndex, 1, oCratesIndex);
+        auto pItem = std::make_unique<TreeItem>(
```

Comment:
The unique pointer only clutters the code here. IMHO the usage of release(); shows that this is just not a unique pointer. 

rows.append(new TreeItem(this, crate.getName(), crate.getId().toVariant());

Is easy to read and shows the "issue" well.